### PR TITLE
Dict\unique_scalar() function to avoid performance penalty hit when using Dict\unique() on large scalar arrays

### PR DESCRIPTION
**Pull request by ghost**
this avoids a rather large performance hit. (#167)

### CODE REVIEW

File: src/Psl/Dict/unique.php [hunk 1]
```diff
@@ -16,6 +16,11 @@
  */
 function unique(iterable $iterable): array
 {
+    
+    if (\is_array($iterable)) {
+        return \array_unique($iterable);
+    }
```
Comment (azjezz) [inline src/Psl/Dict/unique.php]: please import functions instead using `\` ( `use is_array;` and `use array_unique;` )

**Comment by azjezz** [comment]
Ah, this fails when values are not scalar ( which is something that we do support ) 🤔 

how about we add a new function instead ( e.g: `Dict\unique_scalar` )?

**Comment by ghost** [comment]
Gave it a try to `unique_scalar()`.

File: src/Psl/Dict/unique_scalar.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+use function is_array;
+use function array_unique;
+
+/**
+ * Returns a new dict in which each value appears exactly once.
+ *
+ * @psalm-param iterable<array-key, scalar> $iterable
+ *
+ * @psalm-return array<scalar>
```
Comment (azjezz) [inline src/Psl/Dict/unique_scalar.php]: we don't use `@psalm-` prefix, and we always try to keep the return type narrowed as much as possible 😉 

```suggestion
 * @template Tk of array-key
 * @template Tv of scalar
 *
 * @param iterable<Tk, Tv> $iterable
 *
 * @return array<Tk, Tv>
```

File: src/Psl/Dict/unique_scalar.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ */
+function unique_scalar(iterable $iterable): array
+{
+
```
Comment (azjezz) [inline src/Psl/Dict/unique_scalar.php]: extra line.
```suggestion
```

File: src/Psl/Dict/unique.php [hunk 4]
```diff
@@ -16,6 +16,7 @@
  */
 function unique(iterable $iterable): array
 {
+
```
Comment (azjezz) [inline src/Psl/Dict/unique.php]: extra new line
```suggestion
```

File: tests/Psl/Dict/UniqueScalarTest.php [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+    public function testUniqueScalars(): void
+    {
+        $array   = Vec\fill(10, 'foo');
+        $array[] = 'bar';
+
+        $unique = Dict\unique_scalar($array);
+
+        static::assertCount(2, $unique);
+        static::assertSame('foo', Iter\first($unique));
```
Comment (azjezz) [inline tests/Psl/Dict/UniqueScalarTest.php]: we should also add a test case for other Iterators ( use `Iter\Iterator` class ), and IteratorAggregates ( use `Collection\Map` or any other collection )

```suggestion
        static::assertCount(2, $unique);
        static::assertSame(['foo', 'bar'], $unique);
        static::assertSame('foo', Iter\first($unique));
```

**Review by azjezz — COMMENTED** [review]
few comments.

also, please regenerate the documentation using `php docs/documentor.php`

**Comment by ghost** [comment]
Let me know if I can do anything else. I'll be rather busy until for a few hours, but I can come back to this later today or tomorrow. Thanks for all the feedback.

**Review by azjezz — APPROVED** [review]
this is perfect. thanks 👍

**Pull request merged**